### ISSUE TITLE
Promote partial trace (states and density matrices) from Backend to Quantum Information Module

### ISSUE THREAD

**Issue by igres26** [qiboteam/qibo#1168]
There are some features in the Backend that could be promoted into some other modules as they are useful in a variety of situations. Partial traces are one of them. But maybe more are also good to promote.

**Comment by marekgluza** [comment]
Getting Hermitian conjugate came to my mind

```
self.backend.cast( np.matrix(self.backend.to_numpy(operator)).getH()
```
see use case e.g. here
https://github.com/qiboteam/qibo/blob/fdd066f0cfe5a0536f1f6cbfdfa8fe7615686a46/src/qibo/models/dbi/double_bracket.py#L85C27-L86C62

(As a side note, the name is not optimal getHermitianConj comes to mind)

**Comment by marekgluza** [comment]
Is there consensus on the naming/philosophy for the backend  matrix exponentials?

`self.backend.calculate_matrix_exp(-s, H)`

This code computes $e^{-isH}$, i.e. actually doesn't just do an exponential but more, e.g 'calcuate_unitary_matrix'.

For the dbi model we will be working with evolution oracles. For example if you have a `TrotterHamiltonian` object then you could imagine running 
`self.backend.evolution_oracle(-s, H)`
which would e.g. run Trotter-Suzuki and same call could work on hardware and numpy backends.
Right now the numpy backend and hardware are not allowing to write equivalent code while they could because `self.backend.calculate_matrix_exp(-s, H)` works only for one of them.

TLDR; To me taking an exponential to get a numpy unitary should be part of quantum_info or qibojit.

**Comment by renatomello** [comment]
@marekgluza I don't think there's any consensus right now and/or any objection into creating a submodule e.g. `quantum_info.operations` to house all these transformations. 

Keep in mind that most of the backend end functions were created before the existence of the `quantum_info` module, so that is why they are currently where they are.

**Comment by renatomello** [comment]
@stefanopisoni You could use this issue to list the functions you'd like to see moved to the `quantum_info` module too

**Comment by marekgluza** [comment]
> @marekgluza I don't think there's any consensus right now and/or any objection into creating a submodule e.g. `quantum_info.operations` to house all these transformations.
> 
> Keep in mind that most of these backend methods were created before the existence of the `quantum_info` module, so that is why they are currently where they are.

@stefanopisoni @renatomello if you message me or @Sam-XiaoyueLi on matrix then we could join a zoom in SG time zone. We're learning about the code and it would be instructive to us to learn more about the existing functionalities :)
